File: backend/Services/CartaoService.cs
using Microsoft.EntityFrameworkCore;
using FinancasCantinho.Data;
using FinancasCantinho.Models;
using FinancasCantinho.Models.DTOs;

namespace FinancasCantinho.Services;

public interface ICartaoService
{
    Task<IEnumerable<CartaoDto>> GetAllAsync();
    Task<CartaoDto?> GetByIdAsync(int id);
    Task<CartaoDto> CreateAsync(CriarCartaoRequest request);
    Task<CartaoDto?> UpdateAsync(int id, AtualizarCartaoRequest request);
    Task<bool> DeleteAsync(int id);
}

public class CartaoService : ICartaoService
{
    private readonly AppDbContext _context;

    public CartaoService(AppDbContext context)
    {
        _context = context;
    }

    public async Task<IEnumerable<CartaoDto>> GetAllAsync()
    {
        return await _context.Cartoes
            .OrderBy(c => c.Nome)
            .Select(c => new CartaoDto(c.Id, c.Nome, c.Limite, c.Ativo))
            .ToListAsync();
    }

    public async Task<CartaoDto?> GetByIdAsync(int id)
    {
        var cartao = await _context.Cartoes.FindAsync(id);
        if (cartao == null) return null;

        return new CartaoDto(cartao.Id, cartao.Nome, cartao.Limite, cartao.Ativo);
    }

    public async Task<CartaoDto> CreateAsync(CriarCartaoRequest request)
    {
        var cartao = new Cartao
        {
            Nome = request.Nome,
            Limite = request.Limite
        };

        _context.Cartoes.Add(cartao);
        await _context.SaveChangesAsync();

        return new CartaoDto(cartao.Id, cartao.Nome, cartao.Limite, cartao.Ativo);
    }

    public async Task<CartaoDto?> UpdateAsync(int id, AtualizarCartaoRequest request)
    {
        var cartao = await _context.Cartoes.FindAsync(id);
        if (cartao == null) return null;

        cartao.Nome = request.Nome;
        cartao.Limite = request.Limite ?? cartao.Limite;
        cartao.Ativo = request.Ativo;

        await _context.SaveChangesAsync();

        return new CartaoDto(cartao.Id, cartao.Nome, cartao.Limite, cartao.Ativo);
    }

    public async Task<bool> DeleteAsync(int id)
    {
        var cartao = await _context.Cartoes.FindAsync(id);
        if (cartao == null) return false;

        // Soft delete: apenas desativa
        cartao.Ativo = false;
        await _context.SaveChangesAsync();
        return true;
    }
}

File: frontend/src/services/cartaoService.ts
import api from './api';
import { Cartao } from '../types';

export const cartaoService = {
  async getAll(): Promise<Cartao[]> {
    const response = await api.get<Cartao[]>('/api/cartoes');
    return response.data;
  },

  async getById(id: number): Promise<Cartao> {
    const response = await api.get<Cartao>(`/api/cartoes/${id}`);
    return response.data;
  },

  async create(data: { nome: string; limite?: number | null }): Promise<Cartao> {
    const response = await api.post<Cartao>('/api/cartoes', data);
    return response.data;
  },

  async update(id: number, data: { nome: string; limite?: number | null; ativo: boolean }): Promise<Cartao> {
    const response = await api.put<Cartao>(`/api/cartoes/${id}`, data);
    return response.data;
  },

  async delete(id: number): Promise<void> {
    await api.delete(`/api/cartoes/${id}`);
  },
};
